From cd16953179d65687145d001c5c81fd6ceedc9dc4 Mon Sep 17 00:00:00 2001
From: Miaoqing Pan <miaoqing@codeaurora.org>
Date: Tue, 7 Mar 2023 13:07:00 +0800
Subject: [PATCH 04/17] mac80211: use specfied scan delay

Once the scan delay is specified by user, replace the default.

Change-Id: Ifb6b3cfed3e902c4833fe6b90790275c6cf4a6f8
Signed-off-by: Miaoqing Pan <miaoqing@codeaurora.org>
Signed-off-by: Ronghua Zeng <quic_ronghuaz@quicinc.com>
---
 net/mac80211/scan.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/net/mac80211/scan.c b/net/mac80211/scan.c
index 298c386..bd9b593 100644
--- a/net/mac80211/scan.c
+++ b/net/mac80211/scan.c
@@ -897,6 +897,11 @@ static void ieee80211_scan_state_decision(struct ieee80211_local *local,
 	scan_req = rcu_dereference_protected(local->scan_req,
 					     lockdep_is_held(&local->mtx));
 
+	if (scan_req->duration) {
+		local->next_scan_state = SCAN_SET_CHANNEL;
+		goto out;
+	}
+
 	next_chan = scan_req->channels[local->scan_channel_idx];
 
 	/*
@@ -923,7 +928,7 @@ static void ieee80211_scan_state_decision(struct ieee80211_local *local,
 	}
 
 	local->next_scan_state = next_scan_state;
-
+out:
 	*next_delay = 0;
 }
 
@@ -1008,7 +1013,10 @@ set_channel:
 	 */
 	if ((chan->flags & (IEEE80211_CHAN_NO_IR | IEEE80211_CHAN_RADAR)) ||
 	    !scan_req->n_ssids) {
-		*next_delay = IEEE80211_PASSIVE_CHANNEL_TIME;
+		if (scan_req->duration)
+			*next_delay = msecs_to_jiffies(scan_req->duration);
+		else
+			*next_delay = IEEE80211_PASSIVE_CHANNEL_TIME;
 		local->next_scan_state = SCAN_DECISION;
 		return;
 	}
-- 
2.7.4

